src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -506,16 +506,22 @@ private void FinishRequest(EasyRequest completedOperation, int messageResult)
             {
                 VerboseTrace("messageResult: " + messageResult, easy: completedOperation);
 
-                if (completedOperation._responseMessage.StatusCode != HttpStatusCode.Unauthorized && completedOperation._handler.PreAuthenticate)
+                if (completedOperation._responseMessage.StatusCode != HttpStatusCode.Unauthorized)
                 {
-                    ulong availedAuth;
-                    if (Interop.libcurl.curl_easy_getinfo(completedOperation._easyHandle, CURLINFO.CURLINFO_HTTPAUTH_AVAIL, out availedAuth) == CURLcode.CURLE_OK)
+                    if (completedOperation._handler.PreAuthenticate)
                     {
-                        // TODO: fix locking in AddCredentialToCache
-                        completedOperation._handler.AddCredentialToCache(
-                            completedOperation._requestMessage.RequestUri, availedAuth, completedOperation._networkCredential);
+                        ulong availedAuth;
+                        if (Interop.libcurl.curl_easy_getinfo(completedOperation._easyHandle, CURLINFO.CURLINFO_HTTPAUTH_AVAIL, out availedAuth) == CURLcode.CURLE_OK)
+                        {
+                            // TODO: fix locking in AddCredentialToCache
```

Comment:
This is actually not valid anymore. I will look into all the TODOs (there are 9 others)